[22.01] Drop uWSGI from scaling documentation

mvdbeek opened the pull request:
TODO:
- [x] update gravity so it does what I documented here (read stuff from gravity section)
- [x] ~~Create example configurations (and test them ?)~~ Kind of done in the linked gravity PR

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek committed:
WIP: Drop uWSGI from scaling documentation

mvdbeek committed:
Downgrade pre-assignment to the fallback method it is

mvdbeek committed:
Don't reference uwsgi only option in introductory statement

mvdbeek committed:
More progress, describe some basic gunicorn options

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -22,40 +21,21 @@ Just to be clear: increasing the values of `threadpool_workers` in `galaxy.yml`
+appropriate to your Galaxy version.
+
+Gunicorn is able to serve ASGI applications. Galaxy can act as an ASGI web application since release 21.01,
+and we will drop support for being run as a WSGI application, and hence uWSGI compatibility in Galaxy release 22.05.
```
```suggestion
and we will drop support for being run as a WSGI application via uWSGI in Galaxy release 22.05.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -64,135 +44,81 @@ There are multiple deployment strategies for the Galaxy application that you can
-This is the default out-of-the-box configuration as of Galaxy Release 18.01 and will be deprecated in Galaxy Release 21.09
-in favor of running separate processes for handling web requests (gunicorn serving a fastAPI instance), for handling jobs and
-workflows (using the current webless job handling code) and Celery workers for running long-running and/or CPU-intensive tasks.
+This is the default out-of-the-box configuration as of Galaxy Release 22.01.
```
```suggestion
This is the default out-of-the-box configuration.
```

nsoranzo commented on doc/source/admin/scaling.md:
````diff
@@ -64,135 +44,81 @@ There are multiple deployment strategies for the Galaxy application that you can
-   For these scenarios, **uWSGI + Webless** is the recommended deployment strategy.
-```
+If you're migrating Galaxy to 22.01 or newer we recommend you set up the
+**gunicorn + Webless** strategy below.
````
```suggestion
**Gunicorn + Webless** strategy below.
```

mvdbeek committed:
Recommend a proxy

No, really, it's the better option.

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,124 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
+* [Configuring Gunicorn](https://docs.gunicorn.org/en/latest/settings.html)
 
-Note that uWSGI's YAML parser is hand-coded and not actually conformant to the YAML standard. Specifically:
+Note that by default Galaxy will use [gravity](LINK) to create a [supervisor](LINK) configuration that
```
```suggestion
Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration that
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,124 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
-Configuration is performed in the `uwsgi` section of `galaxy.yml`. You will find that the default, if copied from
-`galaxy.yml.sample`, is commented out. The default configuration options are provided to uWSGI on the command line by
-Galaxy's `run.sh` script.
+We will only outline a few of Gunicorn's options. Consult the Gunicorn documentation for more:
```
```suggestion
We will only outline a few of Gunicorn's options, consult the [Gunicorn documentation](https://docs.gunicorn.org/en/latest/settings.html) for more.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,124 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
 
-Galaxy releases prior to 18.01 (or upgraded-to-18.01+ servers which have not migrated their configuration to the YAML
-format) used an INI-format configuration file, `galaxy.ini`.
+* [Configuring Gunicorn](https://docs.gunicorn.org/en/latest/settings.html)
```
```suggestion
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,124 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
+* `gunicorn_extra_args`: You can specify additional arguments to pass to gunicorn here.
 
 Additional options are explained in the [uWSGI Minutiae](#uwsgi-minutiae) below.
 
```
```suggestion
```

nsoranzo commented on doc/source/admin/scaling.md:
````diff
@@ -237,160 +154,124 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
-  # job handling options
 ```
 
 Some of these options warrant explanation:
````
```suggestion
Some of these options deserve explanation:
```

mvdbeek committed:
Describe job handling

mvdbeek committed:
Improve structure, add links

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Fix typo

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -406,8 +345,7 @@ to specific handlers, or to handler tags, as in the following example:
 
-`run.sh` will start the uWSGI process(es), but you will need to start the webless handler processes yourself. This is
-done on the command line like so:
+`run.sh` will start the gunicorn and job handler process(es), but if you are not using run.sh or the generated supervisor setup you will need to start the webless handler processes yourself. This is done on the command line like so:
```
```suggestion
`run.sh` will start the gunicorn and job handler process(es), but if you are not using `run.sh` or the generated supervisor setup you will need to start the webless handler processes yourself. This is done on the command line like so:
```

mvdbeek committed:
Use backticks around run.sh

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Document job handler pool

mvdbeek commented on doc/source/admin/scaling.md:
````diff
@@ -553,10 +366,30 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+    special: 2
+```
+
+?
````
If you have other idea let me know!

natefoo replied:
Maybe if it's an integer it's assumed to be `_default_` and if it's a dict it works as you've shown it here? Also doesn't handle workflow schedulers currently.

mvdbeek replied:
Ah, that is a good point. We should also anticipate scaling the celery workers.
Maybe
```yaml
gravity:
    ...
    process_count:
        job_handler:
           _default_: 10
          special: 2
       workflow_handler: 3
       celery: 2
```
?

mvdbeek committed:
Drop signals section

mvdbeek committed:
Add tentative instructions for setting number of processes in gravity section

mvdbeek committed:
Describe startup and logging

jdavcs commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,121 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
 
-Galaxy releases prior to 18.01 (or upgraded-to-18.01+ servers which have not migrated their configuration to the YAML
-format) used an INI-format configuration file, `galaxy.ini`.
+Configuration is performed in the `gravity` section of `galaxy.yml`. You will find that the default, if copied from
```
This is not setup yet. The [gravity readme](https://github.com/galaxyproject/gravity#configuration) refers to the *galaxy* section of `galaxy.yml` - I assume that's incorrect. ~~In any case, we need to add a `gravity` section to galaxy's config, right?~~ (I see, it can be manually added to `galaxy.yml`)

mvdbeek replied:
Yeah, there's like 50 todos in here

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -566,221 +394,102 @@ $ ./scripts/galaxy-main -c config/galaxy.yml --server-name handler2 --attach-to-
-The value of `stopwaitsecs` should be at least as large as the smallest value of uWSGI's `reload-mercy`,
-`worker-reload-mercy`, and `mule-reload-mercy` options, all of which default to `60`.
+If you want to run your Galaxy server as a persistent service, you can include the `galaxy` script from Galaxy's
+virtualenv in your process manager configuration. You can then continue using the `galaxyctl` command as usual
```
```suggestion
virtualenv in the configuration of your process manager (e.g. systemd). You can then continue using the `galaxyctl` command as usual
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -566,221 +394,102 @@ $ ./scripts/galaxy-main -c config/galaxy.yml --server-name handler2 --attach-to-
-If using the **uWSGI + Webless** strategy, you'll need to addtionally define job handlers to start. There's no simple
-way to activate a virtualenv when using supervisor, but you can simulate the effects by setting `$PATH` and
-`$VIRTUAL_ENV`:
+## Transparent restarts
```
Maybe move this section after `### Systemd` ? Also, do you want it to be under `## Starting and Stopping`? If yes, you need an additional `#`.

mvdbeek replied:
yep, that's a good idea.

mvdbeek replied:
```suggestion
### Transparent restarts
```

mvdbeek committed:
Move transparent starts into starting and stopping

mvdbeek committed:
Mention systemd in starting and stopping section

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek commented on doc/source/admin/scaling.md:
````diff
@@ -553,9 +366,24 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+```yaml
+gravity:
+  ...
+  process_count:
````
Maybe it'd make more sense to control the process_count within their own key,
so 
```
gravity:
  ...
  job_handler:
    _default_: 
      processes: 4
    special: 2
  workflow_handler:
    processes: 2
  celery:
    concurrency: 2
    extra_args: '--debug'
    
```
I think that would look a little better.

natefoo replied:
`processes` as its own key under `job_handler` or `workflow_handler` doesn't distinguish it as "special" from a pool name (tag). How about:

```gravity:
  ...
  job_handler:
    _default_: 
      processes: 4
    special: 2
  workflow_handler: 2
  celery:
    concurrency: 2
    extra_args: '--debug'
```

natefoo replied:
Although how to specify the name template?

mvdbeek replied:
For the name template we could do:

```yaml
  job_handler:
    _default_: 
      processes: 4
     name_template: the_template_{i}
    special: 2
  workflow_handler:
    _default_:
      processes: 2
```
But since the pools are the top level keys here we can't really have a construct to say that a handler should join multiple pools (and/or tags). Maybe we should make the first portion of the handler name the top level key ?

```yaml
gravity:
    handlers:
      handler:
        processes: 2
        name_template: >
          {name}{process}
        pools:
          - job-handler
          - workflow-scheduler
      handler1:
        processes: 1
        pools:
          - job-handler.special
      handler2:
        processes: 1
        pools:
          - job-handler
          - job-handler.special
```
(handler1 and handler2 are overrides here)

would be the equivalent of the old documentation for
> - `handler0` and `handler2` will handle tool executions that are not explicitly mapped to handlers
> - `handler1` and `handler2` will handle tool executions that are mapped to the `special` handler tag

(and handler 1 will only handle special, which we didn't spell out before)

mvdbeek replied:
I've implemented the last example in https://github.com/galaxyproject/gravity/pull/28/commits/a00df1c671fdc01e3fbc42f64a851a45a37a1870

mvdbeek committed:
Drop TODO, it is implemented now

mvdbeek committed:
Adjust pool example and dumb it down a little

There's no "handler2" anymore that handles special and default jobs.
You can of course still set it up exactly as before, that's in the
test case.

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -22,40 +21,21 @@ Just to be clear: increasing the values of `threadpool_workers` in `galaxy.yml`
-* Supports WebSockets, which enable Galaxy Interactive Environments out-of-the-box without a proxy server or Node.js
+as a standalone server with no web stack. Prior to the 18.01 release, Galaxy (by default)
+used the [Python Paste][paste] web stack, and ran in a single process. Between the 18.01 release and the 22.01 release
+uWSGI was used as the default application server. Starting with the 22.01 release the default application server is
```
```suggestion
[uWSGI](https://uwsgi-docs.readthedocs.io/) was used as the default application server. Starting with the 22.01 release the default application server is
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -22,40 +21,21 @@ Just to be clear: increasing the values of `threadpool_workers` in `galaxy.yml`
+as a standalone server with no web stack. Prior to the 18.01 release, Galaxy (by default)
+used the [Python Paste][paste] web stack, and ran in a single process. Between the 18.01 release and the 22.01 release
+uWSGI was used as the default application server. Starting with the 22.01 release the default application server is
+Gunicorn. For information about uWSGI in the Galaxy context please consult the version of this document that is
```
```suggestion
[Gunicorn][Gunicorn]. For information about uWSGI in the Galaxy context please consult the version of this document that is
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -22,40 +21,21 @@ Just to be clear: increasing the values of `threadpool_workers` in `galaxy.yml`
+used the [Python Paste][paste] web stack, and ran in a single process. Between the 18.01 release and the 22.01 release
+uWSGI was used as the default application server. Starting with the 22.01 release the default application server is
+Gunicorn. For information about uWSGI in the Galaxy context please consult the version of this document that is
+appropriate to your Galaxy version.
```
```suggestion
appropriate to your Galaxy release.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -64,135 +44,81 @@ There are multiple deployment strategies for the Galaxy application that you can
 
-Other deployment strategies were commonly used in older versions of Galaxy and can be found in previous versions of the
-documentation, but these are deprecated and should no longer be used.
+If you're migrating Galaxy to 22.01 or newer we recommend you set up the
```
```suggestion
If you're upgrading Galaxy to release 22.01 or newer we recommend you set up the
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -64,135 +44,81 @@ There are multiple deployment strategies for the Galaxy application that you can
-  they are configured. This is the default if no handlers are defined, or handlers are defined but no assign_with attribute is set
-  on the `handlers` tag and *Database SKIP LOCKED* is not available and no uwsgi-mule-messaging is set up.
+  they are configured. `db-transaction-isolation` is the default assignment method if no handlers are defined,
+  or handlers are defined but no assign_with attribute is set on the `handlers` tag and *Database SKIP LOCKED* is not available.
```
```suggestion
  or handlers are defined but no `assign_with` attribute is set on the `handlers` tag and *Database SKIP LOCKED* is not available.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -64,135 +44,81 @@ There are multiple deployment strategies for the Galaxy application that you can
+* **Database SKIP LOCKED** (`db-skip-locked`, new in 19.01) - Jobs are assigned a handler by handlers selecting the unassigned job from
   the database using `SELECT ... FOR UPDATE SKIP LOCKED` on databases that support this query (see the next section for
   details). This occurs via the same process as *Database Transaction Isolation*, the only difference is the way in
   which handlers query the database. This is the default if no handlers are defined, or handlers are defined but no assign_with attribute is set
```
```suggestion
  which handlers query the database. This is the default if no handlers are defined, or handlers are defined but no `assign_with` attribute is set
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -203,30 +129,21 @@ assignment method.
-*Database Transaction Isolation* instead. A detailed explanation of these database locking methods in PostgreSQL can be
+The "database locking" methods (*Database SKIP LOCKED* and *Database Transaction Isolation*) were created to solve
+these issues. The preferred method between the two options is *Database SKIP LOCKED*, but it requires PostgreSQL 9.5
+or newer, sqlite 3.25 or newer or MySQL 8.0 or newer (untested), or MariaDB 10.3 or newer (untested).
```
```suggestion
or newer, SQLite 3.25 or newer or MySQL 8.0 or newer (untested), or MariaDB 10.3 or newer (untested).
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -64,135 +44,81 @@ There are multiple deployment strategies for the Galaxy application that you can
-
-This is the recommended deployment strategy when **Zerg Mode** is used, for Galaxy servers that run web servers and
-job handlers **on different hosts**, and for deployments where dynamic handler addition is desired.
+assigned by the web worker when the job is submitted via the UI/API, meaning that jobs may be assigned to unresponsive handlers.
```
This is covered in much more detail below, drop?

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -203,30 +129,21 @@ assignment method.
+The preferred assignment method is *Database SKIP LOCKED* or *Database Transaction Isolation*.
 
 [2ndquadrant-skip-locked]: https://blog.2ndquadrant.com/what-is-select-skip-locked-for-in-postgresql-9-5/
 [2ndquadrant-blog]: https://blog.2ndquadrant.com/
```
```suggestion
[2ndquadrant-skip-locked]: https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/
[2ndquadrant-blog]: https://www.2ndquadrant.com/en/blog/
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,116 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
-Galaxy releases prior to 18.01 (or upgraded-to-18.01+ servers which have not migrated their configuration to the YAML
-format) used an INI-format configuration file, `galaxy.ini`.
+Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration that
+uses Gunicorn configuration values read from the `gravity` section of your `galaxy.yml` file.
```
```suggestion
Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration
using Gunicorn configuration values read from the `gravity` section of your `galaxy.yml` file.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,116 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
+Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration that
+uses Gunicorn configuration values read from the `gravity` section of your `galaxy.yml` file.
+This is the preferred and out-of-the box way of configuring Gunicorn for serving Galaxy.
+If you are not using `./run.sh` for starting Galaxy or you would like to use another process manager
```
```suggestion
If you are not using `./run.sh` for starting Galaxy or you would like to use another process manager,
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,116 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
+uses Gunicorn configuration values read from the `gravity` section of your `galaxy.yml` file.
+This is the preferred and out-of-the box way of configuring Gunicorn for serving Galaxy.
+If you are not using `./run.sh` for starting Galaxy or you would like to use another process manager
+all of the Gunicorn configuration values can also directly be set on the command line.
```
```suggestion
all the Gunicorn configuration values can also directly be set on the command line.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,116 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
-  example below.
-* Quoting values (with single or double quotes) is unncessary since the parser treats all values as strings. The parser
-  does not correctly handle these quote characters, resulting in invalid values.
+After making changes to the gravity section you always need to activate Galaxy's virtualenv and run `galaxyctl update`
```
```suggestion
After making changes to the `gravity` section, you always need to activate Galaxy's virtualenv and run `galaxyctl update`
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,116 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
-Additional options are explained in the [uWSGI Minutiae](#uwsgi-minutiae) below.
+* `workers`: Controls the number of Galaxy application processes Gunicorn will spawn. Increased web performance can be
+  attained by increasing this value. If Gunicorn is the only application on the server the number of CPUs * 2 + 1 is
+  a good starting value, 4 - 12 workers should be able to handle hundreds if not thousands of requests per second.
```
```suggestion
  attained by increasing this value. If Gunicorn is the only application on the server, a good starting value is the number of CPUs * 2 + 1.
4-12 workers should be able to handle hundreds if not thousands of requests per second.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,116 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
-You can run the startup-time setup steps as the galaxy user after upgrading Galaxy with `sh
-./scripts/common_startup.sh`.
+To bind to ports < 1024 (e.g. if you want to bind to the standard HTTP/HTTPS ports 80/443), you must bind as the `root`
+user and drop privileges to the Galaxy user. However you are strongly encouraged to setup [Apache](apache.md) or [nginx](nginx.md) instead.
```
```suggestion
user and drop privileges to the Galaxy user. However you are strongly encouraged to setup a proxy server
as described in the [production configuration](production.md) documentation.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -553,234 +361,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+          - workflow-scheduler
+      special:
+        pools:
+          - job-handler.special
```
```suggestion
  handlers:
    handler:
      processes: 3
      pools:
        - job-handler
        - workflow-scheduler
    special:
      pools:
        - job-handler.special
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -237,160 +154,116 @@ Handlers (as well as assignment methods) are not configurable when using **uWSGI
 
-Galaxy releases prior to 18.01 (or upgraded-to-18.01+ servers which have not migrated their configuration to the YAML
-format) used an INI-format configuration file, `galaxy.ini`.
+Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration that
```
Where is the supervisor configuration file created?

mvdbeek replied:
You mean the command or the filesystem location ?

nsoranzo replied:
Not sure how this works, I supposed that gravity creates a `supervisord.conf` file somewhere?

mvdbeek replied:
Ah, yes, that's in gravity's state_dir, by default in `database/gravity/supervisor`

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -553,234 +361,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+These commands are available from within Galaxy's virtualenv, or you can install them globally.
+If you have used the standard installation method for Galaxy by running `./run.sh` or executing
+`./scripts/common_startup.sh`, a default directory has been configured in which gravity stores
+it's state. If you have installed Galaxy or gravity by another means you can use the `--state-dir` argument
```
```suggestion
its state. If you have installed Galaxy or gravity by another means, you can use the `--state-dir` argument
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -553,234 +361,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+If you have used the standard installation method for Galaxy by running `./run.sh` or executing
+`./scripts/common_startup.sh`, a default directory has been configured in which gravity stores
+it's state. If you have installed Galaxy or gravity by another means you can use the `--state-dir` argument
+or the `GRAVITY_STATE_DIR` environment variable to control the state dir.
```
```suggestion
or the `GRAVITY_STATE_DIR` environment variable to control the state directory.
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -553,234 +361,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+Galaxy can be started in the background by running `./run.sh --daemon`.
 
-Lastly, collect the tasks defined above into a single group. If you are not using webless handlers this is as simple as:
+Alternatively, you can control galaxy using the `galaxyctl` command provided by gravity.
```
```suggestion
Alternatively, you can control Galaxy using the `galaxyctl` command provided by gravity.
```

nsoranzo commented on doc/source/admin/scaling.md:
````diff
@@ -553,234 +361,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+[2022-02-17 16:31:26,825: DEBUG/MainProcess] beat: Waking up in 5.00 minutes.
+```
+
+More advanced logging options are described in the Galaxy [Logging Configuration documentation](config_logging)
````
```suggestion
More advanced logging options are described in the Galaxy [Logging Configuration documentation](config_logging).
```

nsoranzo commented on doc/source/admin/scaling.md:
```diff
@@ -553,234 +361,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
 
-Sample **uWSGI + Webless** strategy for systemd. More information on systemd.service environment
+This is a sample config for systemd. More information on systemd.service environment
 settings can be found in the [documentation](http://0pointer.de/public/systemd-man/systemd.exec.html#Environment=)
```
```suggestion
settings can be found in the [documentation](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Environment=)
```

mvdbeek committed:
Improve scaling documentation

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Mention supervisor directory and gravity state dir

mvdbeek committed:
Mention that update or one of the start/stop/restart commands need to be run to update the configuration

mvdbeek committed:
Redundant paragraph on why legacy assignment methods are not good

natefoo commented on doc/source/admin/scaling.md:
```diff
@@ -64,135 +44,79 @@ There are multiple deployment strategies for the Galaxy application that you can
 
 * Job handlers and web workers are the same processes and cannot be separated
-* The web worker that receives the job request from the UI/API will be the job handler for that job
+* A web worker that receives the job request from the UI/API will be the job handler for that job
```
```suggestion
* A random web worker will be the job handler for that job
```

natefoo commented on doc/source/admin/scaling.md:
```diff
@@ -553,234 +362,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
+If you have used the standard installation method for Galaxy by running `./run.sh` or executing
+`./scripts/common_startup.sh`, a default directory has been configured in which gravity stores
+its state. If you have installed Galaxy or gravity by another means, you can use the `--state-dir` argument
+or the `GRAVITY_STATE_DIR` environment variable to control the state directory.
```
```suggestion
or the `GRAVITY_STATE_DIR` environment variable to control the state directory. If a state dir is not specified, it defaults to `~/.config/galaxy-gravity`.
```

natefoo commented on doc/source/admin/scaling.md:
```diff
@@ -553,234 +362,151 @@ as explained in [Job Handler Assignment Methods](#job-handler-assignment-methods
 
-Lastly, collect the tasks defined above into a single group. If you are not using webless handlers this is as simple as:
+Alternatively, you can control Galaxy using the `galaxyctl` command provided by gravity.
+After activating Galaxy's virtualenv environment you can start Galaxy in the background
```
```suggestion
After activating Galaxy's virtual environment you can start Galaxy in the background
```

natefoo reviewed (approved):
Thank you for this herculean effort, @mvdbeek!

mvdbeek committed:
Mention default gravity state dir if not running from galaxy's virtualenv

Co-authored-by: Nate Coraor <nate@bx.psu.edu>

mvdbeek committed:
Just say that a random web worker will the the job handler

Co-authored-by: Nate Coraor <nate@bx.psu.edu>
